lib/linked_list.rb

```diff
@@ -12,152 +11,316 @@ def initialize(value, next_node = nil)
+      while current.next != nil
+        if current.next.data == value
+          current.next = current.next.next
+          break
+        end
+        current = current.next
+      end
     end
+  end
 
-    # find the nth node from the end and return its value
-    # assume indexing starts at 0 while counting to n
-    # Time Complexity:  
-    # Space Complexity
-    def find_nth_from_end(n)
-      raise NotImplementedError
+  # method to reverse the singly linked list
+  # note: the nodes should be moved and not just the values in the nodes
+  # Time Complexity: O(n) where n is the number of nodes in the list
+  # Space Complexity: O(1)
+  def reverse
+    previous = nil
+    current = @head
+    after = @head.next
+    while current != nil
+      after = current.next
+      current.next = previous 
+      previous = current
+      current = after
     end
+    @head = previous
+  end
 
-    # checks if the linked list has a cycle. A cycle exists if any node in the
-    # linked list links to a node already visited.
-    # returns true if a cycle is found, false otherwise.
-    # Time Complexity:  
-    # Space Complexity
-    def has_cycle
-      raise NotImplementedError
+  ## Advanced Exercises
+  # returns the value at the middle element in the singly linked list
+  # Time Complexity: O(n) where n is number of nodes in the list
+  # Space Complexity: O(1)
+  def find_middle_value
+    if self.length == 0
+      return nil
     end
 
+    middle = self.length / 2
+
+    middle_value = self.get_at_index(middle)
+    return middle_value
+  end
 
-    # Additional Exercises 
-    # returns the value in the first node
-    # returns nil if the list is empty
-    # Time Complexity:  
-    # Space Complexity
-    def get_first
-      raise NotImplementedError
+  # find the nth node from the end and return its value
+  # assume indexing starts at 0 while counting to n
+  # Time Complexity: O(n) where n is the number of nodes in the list
+  # Space Complexity: O(1)
+  def find_nth_from_end(n)
+    if n >= self.length
+      return nil
     end
 
-    # method that inserts a given value as a new last node in the linked list
-    # Time Complexity:  
-    # Space Complexity
-    def add_last(value)
-      raise NotImplementedError
+    index_of_n = self.length - 1 - n
+    i = 0
+    current = @head
+    until i == index_of_n
+      current = current.next
+      i += 1
     end
+    return current.data
+  end
 
-    # method that returns the value of the last node in the linked list
-    # returns nil if the linked list is empty
-    # Time Complexity:  
-    # Space Complexity
-    def get_last
-      raise NotImplementedError
+  # checks if the linked list has a cycle. A cycle exists if any node in the
+  # linked list links to a node already visited.
+  # returns true if a cycle is found, false otherwise.
+  # Time Complexity: O(n) where n is the number of nodes in the list
+  # Space Complexity: O(1)
+  def has_cycle
+    if self.length <= 1

```

Comment:
Just a point, if it has a cycle, then the length method will get stuck in an infinite loop.